Fix the visibility of partially out of stock variable products when using the layered nav widget

Konamiman opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR fixes an issue that appears when using the layered nav widget (which allows filtering the listed products by an attribute) and there are variable products where some variations have stock and others don't. The issue is that when an attribute is selected for filtering, all variable products having a variation that corresponds to that attribute will be displayed, even those not having stock for that variation (as long as at least one variation has stock).

Example: you sell shoes in two colors, black and blue. Black is out of stock but there's stock for blue. You select "black" (but not "blue") in the widget. The shoes will be included in the products list, but they shouldn't.

Additionally, the count of products that appear in the widget for each attribute is also wrongly including the variable products that don't have stock for the attribute corresponding to that variation.

Note that the original issue is about products out of stock, but the changes act actually on the more wide visibility condition. A product is "visible" if it's published, not excluded for catalog, and has stock. Additionally, a variable product will not be considered visible if the parent product is not.

#### New behavior of the counters

Now the counters count all the variations that are in stock, regardless of whether the filter uses OR or AND logic; that's because one given variation has always one single value for the attribute, so the logic in use is irrelevant. For non-variable products they continue working as usual.

Variation counts will be equivalent to product counts when one single attribute is used to define variations, but if more than one attribute is used that could not be the case. For example you have a "shoes" product with two attributes used for variations, "color" and "style"; if you define variations as "black classic" and "black sport" then "black" will be counted twice in the filter (if both variations have stock).

Also, when using multiple filters they interact with AND logic as usual: only variations that have both attributes in its definition are counted.

Closes #25524.

### How to test the changes in this Pull Request:

**NOTE:** You need to run database updates for the widget counters to work properly!

1. Go to Settings > Products > Inventory, mark "Hide out of stock items from the catalog".
2. Create an attribute named "Color", with the following terms: Black, Brown, Blue, Green, Pink.
3. Create the following variable products: Big shoes, Medium shoes, Small shoes, Kid shoes. Create variations for all the existing colors on all of them. Assign any regular price to all the variations. (Hint: create one product, then duplicate)
4. Modify the products so the following variations are in stock and the others are out of stock:
* Big: black, brown.
* Medium: blue, brown.
* Small: blue, green.
* Kids: green, pink.
5. Add the "Filter by attribute" widget, filtering by color with OR.
6. Try filtering by one or two colors and verify that the results make sense. For example when filtering by "blue" you should see only "Medium shoes" and "Small shoes" (without the fix you would see all four shoe types). Verify also that the results count that appear at the beginning and the end of the list is correct.
7. Change the widget filtering type to AND and try different combinations again. Verify that selecting pink and black displays the "No products available" message.

#### How to test the counters

With the same setup, verify that the counters display the following, no matter what attributes you select in the widget:

![image](https://user-images.githubusercontent.com/937723/81082993-4b04ed80-8ef4-11ea-8d0d-6e8c5a14d432.png)

Now make sure that nothing is selected, then go and change the widget operator to AND. Reload the products page and verify that the behavior is the same.

#### Testing counters with multiple attributes

1. Create an attribute named "Style" with two terms: Classic, and Sport.
2. Modify all the products so that all of them use the new style for variations, but leave the value for all variations as "Any style".
3. Modify the Big shoes product so that the Black variation is Classic, and create a new variation that is Black Sport (leave both in stock).
4. Modify the Medium shoes product so that the Brown variation is Sport (and still has stock).
5. Add the "Filter by attribute" widget, filtering by style with any logic.
6. Reload the page and verify that the counters now appear like this:

![image](https://user-images.githubusercontent.com/937723/87396534-42dfc300-c5b3-11ea-8bb0-7f040e5c2206.png)

Note how there are two variations listed for Black, since we added a new one to the Big shoes. Also note how the two classic variations and the single sport variation are listed in the style filter.

7. Select "Black" in the color filter and note how the style filter now displays a value of 1 for classic and 1 for sport, since there's one visible variation of each.
8. Unselect all and select "Blue" in the color filter and note how the style filter disappears, since there's no variation that is blue and has a style associated.
9. Unselect all in the color filter and select "Sport" in the style filter, you will see this:

![image](https://user-images.githubusercontent.com/937723/87396197-c1883080-c5b2-11ea-9720-7e86fe4d298f.png)

#### Testing visibility with multiple filters

While testing counters with multiple filters you may have noticed that regardless of what filters you selected in the style filter you were still seeing all the products in the catalog (if you don't filter by color). That's because all of the products have visible variations with "Any style" and thus they always match the style filter.

Modify the "Kid shoes" product so that all the variations have the sport style, then select "Classic" in the style filter while not filtering by color, and you'll see that the product disappears from the catalog.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - The filtering widget now works as expected with variable products, displaying those products for which visible variations are available.

Konamiman committed:
Fix code sniffer errors in some files.

Fixed files:

includes/abstracts/abstract-wc-product.php
includes/class-wc-query.php
includes/wc-template-functions.php
includes/widgets/class-wc-widget-layered-nav.php
templates/loop/result-count.php
tests/unit-tests/util/class-wc-tests-wc-query.php

Konamiman committed:
Fix total products count display when there's one single result.

Fix from "Showing all 1 results" to "Showing the single result".

Konamiman committed:
Fix visibility of variable products with layered nav filtering.

The layered nav filtering doesn't work well with variable products
when some variations have stock and other don't. When a term is
selected in the widget, a variable product having no stock for
the variation corresponding to that term but having stock for
other variations will be displayed, but it shouldn't.

This commit fixes that by introducing two changes:

- A new override of "is_visible" for WC_Product_Variable that
  looks at the supplied filters, compares them against the corresponding
  available variations and calculates the visibility based on
  the query type (OR or AND).

- A hook on the "found_posts" filter in WC_Query, that adjusts
  the posts count based on the found products visibility
  when there are filters available; this is needed to sync the
  "displaying X posts" messages and the paging when variable
  products are hidden due to stock status.

Additionally, the visibility calculated in "found_posts" is cached
as loop variables so that it isn't calculated again when actually
displaying the products.

Konamiman committed:
Fix code sniffer errors in some files.

Fixed files:

includes/data-stores/class-wc-product-variable-data-store-cpt.php
includes/wc-update-functions.php

Konamiman committed:
Refactor WC_Helper_Product by extracting duplicated code to methods.

The new methods are create_product_variation_object and
create_product_attribute_object, they are public.

Konamiman committed:
Record attribute terms for product variations in wp_term_relationships.

Product attributes are currently recorded as terms in
wp_term_relationships (product attributes are actually taxonomies).
In the case of variable products this is true for the main product,
but not for the variations. The attributes used to define variations
are stored as post meta, but nothing is recorded in the term
relationships table.

This is a problem when using the layered nav filtering plugin,
since the attribute counters displayed are calculated based solely
on the contents of the term relationships table. Adding meta queries
would be really messy (especially when the widget is configured
with AND operator) and would probably also hurt performance.

This commit adds a change to store the attributes for variations
as term relationships, additionally to storing them as post meta.
Terms are stored on variation creation, and updated/deleted together
with the variation as appropriate. "Any" variations (stored in meta
as empty values) are not stored as terms.

Additionally, a database upgrade is included in order to backfill
terms for already existing products.

Konamiman committed:
Fix counters in nav filtering widgets for variable products.

After the change that registers variation attributes as terms
(in addition to reigstering them as post meta) it is now time
to modify the get_filtered_term_product_counts methods in
WC_Widget_Layered_Nav so that it works consistently for both
variable and non-variable products. The logic for the counters
is now as follows:

with OR operator:
- Simple products: count the attributes of all visible products
  (unchanged behavior).
- Variable products: count attributes corresponding to
  visible variations.

with AND operator:
- Simple products: count the attributes of visible products but only
  for products that have all the selected (unchanged behavior).
- Variable products: find all the products for which all the variations
  corresponding to the selected attributes exist and are visible,
  then count the attributes corresponding to the visible variations
  of those products.

A product is "visible" if it's published, not excluded for catalog,
and has stock. Additionally, a variable product will not be considered
visible if the parent product is not.

claudiosanches commented on includes/abstracts/abstract-wc-product.php:
```diff
@@ -1963,7 +1974,8 @@ public function get_file_download_path( $download_id ) {
 		$html = '';
 
-		if ( ( $suffix = get_option( 'woocommerce_price_display_suffix' ) ) && wc_tax_enabled() && 'taxable' === $this->get_tax_status() ) { // @phpcs:ignore Squiz.PHP.DisallowMultipleAssignments.Found, WordPress.CodeAnalysis.AssignmentInCondition.Found
+		// phpcs:ignore Squiz.PHP.DisallowMultipleAssignments.FoundInControlStructure, WordPress.CodeAnalysis.AssignmentInCondition.Found
```
You can remove both lines and fix the conditional.

Konamiman replied:
Ok, thanks for spotting those. I'll wait in case someone else (or yourself) comes up with more, then fix all of them in a single refactoring commit.

claudiosanches commented on includes/abstracts/abstract-wc-product.php:
```diff
@@ -1871,7 +1881,8 @@ public function get_image( $size = 'woocommerce_thumbnail', $attr = array(), $pl
 	 */
 	public function get_shipping_class() {
-		if ( $class_id = $this->get_shipping_class_id() ) { // @phpcs:ignore Squiz.PHP.DisallowMultipleAssignments.Found, WordPress.CodeAnalysis.AssignmentInCondition.Found
+		// phpcs:ignore Squiz.PHP.DisallowMultipleAssignments.FoundInControlStructure, WordPress.CodeAnalysis.AssignmentInCondition.Found
```
Better stop ignoring those rules, we just need to fix the `if` moving the assignment to a variable.

claudiosanches reviewed (changes requested):
I still need to test it in full, but here's some coding standards that can be fixed without excluding from PHPCS.

Konamiman committed:
Adjustments in WC_Query::adjust_posts_count

Two adjustments were needed:

- Adjust the count even when there's no nav filtering in the query.
  This is necessary to present the proper products count.
  even when the woocommerce_product_is_visible filter is used.
- Account for the case where $GLOBALS['wp_query']->posts
  returns objects instead of ids (for example when viewing
  a product page).

Konamiman committed:
Remove two phpcs:ignores, improve code instead

Konamiman committed:
Fix: WC_Query()::adjust_posts_count failing when posts list is null.

claudiosanches commented on includes/class-wc-product-variation.php:
```diff
@@ -580,4 +583,22 @@ protected function get_valid_tax_classes() {
+	/**
+	 * Delete variation, set the ID to 0, and return result.
+	 *
+	 * @since  2.6.0
```
```suggestion
	 * @since  4.4.0
```

claudiosanches commented on includes/class-wc-install.php:
```diff
@@ -148,6 +148,10 @@ class WC_Install {
 			'wc_update_400_reset_action_scheduler_migration_status',
 			'wc_update_400_db_version',
 		),
+		'4.2.0' => array(
```
```suggestion
		'4.4.0' => array(
```

claudiosanches commented on includes/class-wc-install.php:
```diff
@@ -148,6 +148,10 @@ class WC_Install {
 			'wc_update_400_db_version',
 		),
+		'4.2.0' => array(
+			'wc_update_420_insert_attribute_terms_for_variable_products',
```
```suggestion
			'wc_update_440_insert_attribute_terms_for_variable_products',
```

claudiosanches commented on includes/class-wc-install.php:
```diff
@@ -148,6 +148,10 @@ class WC_Install {
 		),
+		'4.2.0' => array(
+			'wc_update_420_insert_attribute_terms_for_variable_products',
+			'wc_update_420_db_version',
```
```suggestion
			'wc_update_440_db_version',
```

claudiosanches commented on includes/class-wc-query.php:
```diff
@@ -363,6 +367,60 @@ public function remove_product_query_filters( $posts ) {
+	 * We do that by just checking if each product is visible.
+	 *
+	 * We also cache the post visibility so that it isn't checked again when displaying the posts list.
+	 *
```
```suggestion
	 *
	 * @since 4.4.0
```

claudiosanches commented on includes/class-wc-query.php:
```diff
@@ -666,9 +728,11 @@ public function get_tax_query( $tax_query = array(), $main_query = false ) {
 			$product_visibility_not_in[] = $product_visibility_terms['outofstock'];
 		}
 
+		// phpcs:disable WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.MissingUnslash
```
```suggestion
		// phpcs:disable WordPress.Security.NonceVerification.Recommended
```

claudiosanches commented on includes/class-wc-query.php:
```diff
@@ -685,6 +749,7 @@ public function get_tax_query( $tax_query = array(), $main_query = false ) {
 				);
 			}
 		}
+		// phpcs:enable WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.MissingUnslash
```
```suggestion
		// phpcs:enable WordPress.Security.NonceVerification.Recommended
```

claudiosanches commented on includes/class-wc-query.php:
```diff
@@ -666,9 +728,11 @@ public function get_tax_query( $tax_query = array(), $main_query = false ) {
-			$rating_filter = array_filter( array_map( 'absint', explode( ',', $_GET['rating_filter'] ) ) ); // WPCS: input var ok, CSRF ok, Sanitization ok.
+		if ( isset( $_GET['rating_filter'] ) ) {
+			// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+			$rating_filter = array_filter( array_map( 'absint', explode( ',', $_GET['rating_filter'] ) ) );
```
```suggestion
			$rating_filter = array_filter( array_map( 'absint', explode( ',', wp_unslash( $_GET['rating_filter'] ) ) ) );
```

claudiosanches commented on includes/data-stores/class-wc-product-variable-data-store-cpt.php:
```diff
@@ -200,7 +200,7 @@ public function read_variation_attributes( &$product ) {
 				}
 
-				// Empty value indicates that all options for given attribute are available.
+				// Empty value inicates that all options for given attribute are available.
```
Sounds like you introduced a typo here.

claudiosanches commented on includes/wc-template-functions.php:
```diff
@@ -8,6 +8,8 @@
  * @version  2.5.0
  */
 
+// phpcs:disable Generic.Commenting.Todo.TaskFound
```
You can remove it, we should get notices about TODOs.

claudiosanches commented on includes/wc-template-functions.php:
```diff
@@ -232,6 +238,27 @@ function wc_set_loop_prop( $prop, $value = '' ) {
 
+/**
+ * Set the current visbility for a product in the woocommerce_loop global.
+ *
```
```suggestion
 *
 * @since 4.4.0
```

claudiosanches commented on includes/wc-template-functions.php:
```diff
@@ -232,6 +238,27 @@ function wc_set_loop_prop( $prop, $value = '' ) {
+
+/**
+ * Gets the cached current visibility for a product from the woocommerce_loop global.
+ *
```
```suggestion
 *
 * @since 4.4.0
```

claudiosanches commented on includes/wc-template-functions.php:
```diff
@@ -241,7 +268,7 @@ function wc_set_loop_prop( $prop, $value = '' ) {
  */
 function woocommerce_product_loop() {
-	return have_posts() || 'products' !== woocommerce_get_loop_display_mode();
+	return wc_get_loop_prop( 'total' ) > 0 || 'products' !== woocommerce_get_loop_display_mode();
```
`wc_get_loop_prop()` it set again the loop, also `have_posts()` works different, and probably better keep it for backwards compatibility.

Konamiman replied:
I don't remember why I made this change, maybe I needed it while I was in the middle of the development; but I have verified that everything still works after reverting to using `have_posts` again so I did and pushed that.

claudiosanches commented on includes/wc-template-functions.php:
```diff
@@ -704,7 +731,9 @@ function wc_product_class( $class = '', $product_id = null ) {
 	if ( is_null( $values ) ) {
-		$values = $_GET; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		// phpcs:disable WordPress.Security.NonceVerification.Recommended
+		$values = $_GET;
```
It should work with just one line using `phpcs:ignore WordPress.Security.NonceVerification.Recommended`

Konamiman replied:
I could swear that I did it that way because it wasn't accepting the one-liner version. But now it is accepting it. 🤷🏻‍♂️

claudiosanches commented on includes/wc-update-functions.php:
```diff
@@ -2110,3 +2130,48 @@ class_exists( 'ActionScheduler_DataController' ) &&
+ *
+ * @return bool true if there are more products to process.
+ */
+function wc_update_420_insert_attribute_terms_for_variable_products() {
```
```suggestion
function wc_update_440_insert_attribute_terms_for_variable_products() {
```

claudiosanches commented on includes/wc-update-functions.php:
```diff
@@ -2110,3 +2130,48 @@ class_exists( 'ActionScheduler_DataController' ) &&
+/**
+ * Update DB version.
+ */
+function wc_update_420_db_version() {
```
```suggestion
function wc_update_440_db_version() {
```

claudiosanches commented on includes/wc-update-functions.php:
```diff
@@ -2110,3 +2130,48 @@ class_exists( 'ActionScheduler_DataController' ) &&
+ * Update DB version.
+ */
+function wc_update_420_db_version() {
+	WC_Install::update_db_version( '4.2.0' );
```
```suggestion
	WC_Install::update_db_version( '4.4.0' );
```

claudiosanches commented on includes/widgets/class-wc-widget-layered-nav.php:
```diff
@@ -397,17 +433,119 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
 
+	/**
+	 * Wrapper for WC_Query::get_main_tax_query() to ease unit testing.
+	 *
```
```suggestion
	 *
	 * @since 4.4.0
```

claudiosanches commented on includes/widgets/class-wc-widget-layered-nav.php:
```diff
@@ -397,17 +433,119 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
+
+	/**
+	 * Wrapper for WC_Query::get_main_search_query_sql() to ease unit testing.
+	 *
```
```suggestion
	 *
	 * @since 4.4.0
```

claudiosanches commented on includes/widgets/class-wc-widget-layered-nav.php:
```diff
@@ -397,17 +433,119 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
+
+	/**
+	 * Wrapper for WC_Query::get_main_search_queryget_main_meta_query to ease unit testing.
+	 *
```
```suggestion
	 *
	 * @since 4.4.0
```

claudiosanches commented on includes/widgets/class-wc-widget-layered-nav.php:
```diff
@@ -397,17 +433,119 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
+	/**
+	 * Get a tax query SQL for a given set of taxonomy, terms and operator.
+	 * Uses an intermediate WP_Tax_Query object.
+	 *
```
```suggestion
	 *
	 * @since 4.4.0
```

claudiosanches commented on includes/widgets/class-wc-widget-layered-nav.php:
```diff
@@ -397,17 +433,119 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
+
+	/**
+	 * Convert a tax query array to SQL using an intermediate WP_Tax_Query object.
+	 *
```
```suggestion
	 *
	 * @since 4.4.0
```

claudiosanches reviewed (changes requested):
I left some code suggestions for you.

MohsinQK commented:
@Konamiman  I found a strange problem. the filter is working but a bit buggy.
I have one test Product with the Quantities as shown in the screenshot.
![image](https://user-images.githubusercontent.com/1622975/86498952-ca525880-bda1-11ea-88f9-1efe4d432221.png)

in frontend the filter works but buggy for example

1. if I select Green Color (Only) from the color Filter the Product disappears as expected because it's not available in Green Color. (✅)
2.  if I select Large Size (Only) Product appears in the list even its not available in Large Size.(❌)
3. if I select Green Color First **AND** then select Large Size Product not get listed Which is correct.(✅) 
4. if I select Large Size First **AND** then select Green Color Product gets listed which is wrong(❌)

below are all screenshots of the use cases in the order mentioned above in each use case Product must **NOT** appear in the list but in 2 cases it's listing even its Quantity is zero.

This is the Product I am testing on.
![image](https://user-images.githubusercontent.com/1622975/86499309-f8389c80-bda3-11ea-8eaf-369bff6ea6ba.png)

1. 
![image](https://user-images.githubusercontent.com/1622975/86499339-2c13c200-bda4-11ea-9e03-161a12eb4cbf.png)
2. 
![image](https://user-images.githubusercontent.com/1622975/86499379-62514180-bda4-11ea-9005-58a1b2cd2874.png)
3. 
![image](https://user-images.githubusercontent.com/1622975/86499454-c07e2480-bda4-11ea-8867-e807888188ca.png)
4. 
![image](https://user-images.githubusercontent.com/1622975/86499482-eacfe200-bda4-11ea-8c29-fa9e0ff699b1.png)

I think its only working in the Order their Variation attributes are selected from the backend.
Please let me know for any other information you need

Konamiman commented:
Hi @MohsinQK. Regarding your feedback, I assume you are using two separate filtering widgets, one for color and another one for size, at the same time.

> 2. if I select Large Size (Only) Product appears in the list even its not available in Large Size.(❌)

The product is still available in large size, for the Black and Indigo colors. Therefore it should still be listed, right? Please tell me if I'm missing something else.

> 3. if I select Green Color First AND then select Large Size Product not get listed Which is correct.(✅)
> 4. if I select Large Size First AND then select Green Color Product gets listed which is wrong(❌)

This is indeed inconsistent. I'm not sure if the product should be listed or not in these cases (we have "AND"/"OR" behavior selection for each individual widget, but when using two filters at the same time I don't know how it's supposed to behave), but what's for sure is that it should act in the same way in both cases.

I'll have to take a look at this and check with my team, thanks for reporting!

IlPicasso commented:
> Hi @MohsinQK. Regarding your feedback, I assume you are using two separate filtering widgets, one for color and another one for size, at the same time.
> 
> 
> 
> > 1. if I select Green Color First AND then select Large Size Product not get listed Which is correct.(✅)
> 
> This is indeed inconsistent. I'm not sure if the product should be listed or not in these cases (we have "AND"/"OR" behavior selection for each individual widget, but when using two filters at the same time I don't know how it's supposed to behave), but what's for sure is that it should act in the same way in both cases.
> 
> I'll have to take a look at this and check with my team, thanks for reporting!

> > 1. if I select Large Size (Only) Product appears in the list even its not available in Large Size.(❌)
> The product is still available in large size, for the Black and Indigo colors. Therefore it should still be listed, right? Please tell me if I'm missing something else.

I believe this is the correct behaviour. 


> > 2. if I select Large Size First AND then select Green Color Product gets listed which is wrong(❌)

If I choose Large and Green, it is supposed that I am searching for an item that HAS Large Size and green color so everything else should dissapear if there is not those attributes, otherwise I wouldnt filter it that way

Konamiman committed:
Change in the logic of is_visible_core for variable products

Now, if there are filters present the logic is as follows:

- For multiple filtering values of the same attribute:
  the product is visible if there's at least one variation
  that has one of the filtering values associated to the attribute,
  or if there's at least one variation having the attribute
  with a value of "Any".

- For filtering by more than one attribute:
  the product is visible if there's at least one variation that
  is visible for ALL the attributes according to the above rule.

Note that this is irrespective of the type of logic configured for
the filter (OR or AND).

Konamiman committed:
Apply suggestions from code review

Mostly adding "since 4.4.0" annotations and updating
db update functions from 4.2 to 4.4.

Co-authored-by: Claudio Sanches <contato@claudiosanches.com>

Konamiman committed:
Adjustments in the items count calculation on the nav filtering widget.

The calculations are now consistent with the change made to the
visibility of the variable products when using one or multiple filters.

Konamiman commented:
Hi @MohsinQK and @IlPicasso, I have made changes so that:

* The special behavior for the AND filtering case has been removed. One doesn't usually think "I want to see shoes that have stock of blue AND have stock of green" so that logic was unnecessarily complex and confusing.  Now the filters have always OR behavior as far as variations are concerned, since one given variation will always have one single value for a given attribute.
* Filtering by multiple attributes works as expected, regardless of the order in which you select them.

I hope those changes are sensible and useful.

Konamiman committed:
Revert woocommerce_product_loop to use have_posts instead of wc_get_loop_prop( 'total' ).

MohsinQK commented:
@Konamiman yes the Point 2 is correct it's my misunderstanding. but point 4 was not fine for which you made some changes already.

claudiosanches reviewed (approved):
Great work!

jonathansadowski reviewed (approved):
Works well, nice work 👍

Konamiman committed:
Fixes after rebasing.

- Fix a call to get_available_variations with incorrect optional arguments.
- Add code that has been removed in WC_Product_Variable::get_available_variations.

B-Anders commented:
I've noticed a bunch of other strange behavior from that filter. I hope @Konamiman can take a look and make some sense.

To reproduce:

1. Start with one simple product (Product1). It has 2 attributes, "ARTIST" and "COLOR".
ARTIST is used in the attribute filter widget and everything works as expected.
2. Change product to variable and generate two (or more) variations with (only) the second attribute "COLOR". Give it some quantity to have the variations on stock.
3. Check: product shows on the shop and category pages, variations are correct on the single product page.
_Problem: the filter for attribute ARTIST doesn't count the product and the whole widget isn't showing on any pages._
4. Add a second product (Product2): simple, with attribute ARTIST, same value as the first product.
Now the filter widget is displayed again and filters both products correctly. _So now it's counting the variable product too?!?_
5. Edit the variable "Product1" and make variations with the attribute ARTIST and COLOR so that all variations have both attributes (in my case it doesn't make sense to show variation options for ARTIST, but we need this step).
Now the filter is showing on pages as expected even if we delete Product1 > filter is counting the ARTIST attribute.
6. Edit the Product1 again, by unchecking attribute ARTIST from 'use for variations'. Don't touch anything else, just save. The product still has the variations with the COLOR attribute.
_Now the filter is counting the product with ARTIST attribute correctly!!!_
7. ...but the last step only works for the Shop page. The widget isn't showing on the product's category page.

If you know a fix, please let me know, I have just three days left on this project.

Konamiman commented:
@B-Anders (and everybody having similar issues), please take a look at this: https://github.com/woocommerce/woocommerce/issues/27419#issuecomment-689657524

TheGranadilla commented:
Hi,
I was wondering if you were still planning to improve nav filtering by attributes for variables products.
Just to hide results for not in stock items. 

I know it's a complex thing, that is discuss since months, maybe years.
But it's also something important. Actually, the filtering return out of stock variations, which is verry not great for customers.
They filter by a product size, and then, in my case, 50% of the products are out of stock in the variation they were looking for...

The only solution would be to delete the variations, but this is going to be a nightmare, especially since those variations might get back in stock one day... 

Is fixing this still an objective?

Thanks !

Konamiman commented:
Hi @TheGranadilla,

Yes, we plan to take on this problem again at some point in the future. Please see [this comment in the original issue](https://github.com/woocommerce/woocommerce/issues/25524#issuecomment-698869326).

MohsinQK commented:
This feature stopped working again in the latest releases of woocommerce.
in 4.4 it works as expected, but in 4.7 it's not working as expected.
this is a very important feature. is this removed in the latest version or its by mistake?

WooCommerce 4.7
![image](https://user-images.githubusercontent.com/1622975/99149367-61fc7780-26af-11eb-95e9-da7ba4a6efb3.png)

Just Changed the woocommerce version to 4.4 and it works again. out of stock(Red Background) is not in the listing any more.
![image](https://user-images.githubusercontent.com/1622975/99149402-9b34e780-26af-11eb-8649-e9c66af5232e.png)

TheGranadilla commented:
@MohsinQK This was reverted in 4.5 because it introduced new bugs.
But it seems they aren't working anymore on variables products Filtering........
@Konamiman Any update on this? Woocmmerce widget still return out of stock variations....
